### PR TITLE
Mm 19333 migrate sql channel store to use squirrel

### PR DESCRIPTION
**Pull request by src-r-r**
#### Summary

This fixes [19333](https://github.com/mattermost/mattermost-server/issues/19333). The query functions have been converted to squirrel database queries. I verified the relevant tests pass with `make test`. (As of master `9dcd61c12`, several of the tests fail, but these failing tests match this PR's failed tests).

#### Ticket Link

https://github.com/mattermost/mattermost-server/issues/19333

#### Release Note

```release-note
channel store queries have been migrated from sql to squirrel structs.
```

### CODE REVIEW

**Comment by src-r-r** [comment]
This one uses the main repo history instead of https://github.com/mattermost/mattermost-server/pull/19540

**Comment by src-r-r** [comment]
This one uses the main repo history instead of https://github.com/mattermost/mattermost-server/pull/19540

**Comment by agnivade** [comment]
Thank you @src-r-r - Could you please fix the golangci failures? And then we can get started with the PR review.

**Comment by agnivade** [comment]
Thanks @src-r-r -  can you revert the changes to api4/websocket_norace_test.go, and use `err2` as the variable in L3466 to fix the lint issue?

**Comment by src-r-r** [comment]
> Thanks @src-r-r - can you revert the changes to api4/websocket_norace_test.go, and use `err2` as the variable in L3466 to fix the lint issue?

I can. For me running `make golangci-lint` produces this error:

```
api4/websocket_norace_test.go:1: File is not `gofmt`-ed with `-s` (gofmt)
// +build !race
make: *** [Makefile:249: golangci-lint] Error 1
```

So that's why I changed the file. Is this error expected? I'll just trust that the main pipeline won't fail the test.

**Comment by agnivade** [comment]
This happens with Go versions >1.16. We are still on 1.16. It's possible to add that line, but we need to do it in a way that does not break the license check. If you want, you are most welcome to send the fix in a separate PR, but I'd prefer to contain only related changes in this PR.

**Comment by src-r-r** [comment]
Not sure why the tests are suddenly failing. I decided to rebase this branch with the latest master this morning.

**Comment by src-r-r** [comment]
@agnivade, looks like all the tests have passed. Sorry for the false alarms earlier.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3422]: I believe this can be moved back into the previous statement like it was originally.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3428]: Don't think this will be needed after we change the `channels` variable

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3436]: Please return the error upwards.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3433]: Let's avoid named returns please.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3439]: commented code.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3452]: `channels := model.ChannelListWithTeamData{}`

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3457]: let's remove this.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3462]: This was possibly done during debugging. Let's revert.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3468]: Don't think this is needed.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3640]: Please remove this log line.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3656]: Let's rewrite the old method itself rather than adding a separate one?

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3644]: `groupChannels := model.ChannelList{}`

**Review by agnivade — COMMENTED** [review]
Gave an initial first pass. This is excellent work @src-r-r !

**Comment by agnivade** [comment]
@src-r-r - Is this ready for review? I see some TODO comments left in the latest changes.

**Comment by src-r-r** [comment]
@agnivade Thanks. Removed my TODO comments. The changes are now in the pipeline.

**Comment by src-r-r** [comment]
@agnivade Looks like it's just waiting for approval now.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3011]: This is still using the old pattern. Let's change it to the new style as well.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3023]: Why is this needed? We aren't doing it in `AutocompleteInTeam`

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3029]: Same as earlier. Let's use the new style. And inline some of the sub-queries for better readability. And use model.ChannelTypePrivate instead of "P"

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3082]: Use `s.getQueryBuilder().Select(...`

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3104]: This can be attached to the original clause itself.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3281]: Same question as earlier.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3269]: Same as above. Let's use the new style.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3624]: Please update the original methods themselves. There's no point keeping unused code lying around.

File: store/sqlstore/channel_store.go
Comment (agnivade) [inline store/sqlstore/channel_store.go:3563]: Let's update the original method `getSearchGroupChannelsQuery` itself.

**Review by agnivade — COMMENTED** [review]
We are getting close @src-r-r. I think not all the methods have been migrated to the new style, and we need to replace the methods like `getSearchGroupChannelsQuery` with the squirrel code rather than add new ones.

File: store/sqlstore/channel_store.go
Comment (src-r-r) [reply store/sqlstore/channel_store.go:2878]: What do you mean by a nil check?

File: store/sqlstore/channel_store.go
Comment (src-r-r) [reply store/sqlstore/channel_store.go:3624]: Thought I removed them all. Thanks for catching that.

File: store/sqlstore/channel_store.go
Comment (agnivade) [reply store/sqlstore/channel_store.go:2872]: `s.searchClause` returns `nil, nil` if likeClause is nil. We handle this in the previous method by checking for `searchClause == nil`. That's what I mean by a nil check.

**Comment by agnivade** [comment]
@src-r-r - Some of my comments are still not addressed. Please spend some time and take a look at all of my comments. A lot of similar issues are repeated throughout the PR.

**Comment by agnivade** [comment]
@noxer - Could you give this a quick review and see if everything looks good? Thanks.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:2867]: We are using `query` in the rest of the methods. Please change `q` to `query`.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:2910]: You should use `s.getQueryBuilder()` here to set the appropriate placeholders.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3098]: Use `s.getQueryBuilder()`

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3471]: You can replace the `if err != nil` and this return with just the `return channels, errors.Wrapf(...)`. You are returning `channels` in any case and `errors.Wrapf` returns `nil` if `err` is `nil` anyway.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3536]: Just use `sq.Eq` here. You can specify multiple pairs which will be `AND`ed automatically.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3490]: I know this was in the original code, could you still change it to just use `strings.ToLower` and `strings.Fields`, please?

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3496]: Please use `s.getSubQueryBuilder()`.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3511]: Same as above

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3532]: `s.getSubQueryBuilder()`

**Review by noxer — CHANGES_REQUESTED** [review]
Thank you for your work @src-r-r. I've added a few comments but the overall PR is impressive.

File: store/sqlstore/channel_store.go
Comment (agnivade) [reply store/sqlstore/channel_store.go:3471]: Yeah, the end result is the same, but this will make it slightly inconsistent with the rest of the codebase. Let's keep using the same pattern? I have changed it to the short syntax btw.

**Comment by agnivade** [comment]
@noxer - PTAL :)

File: store/sqlstore/channel_store.go
Comment (noxer) [reply store/sqlstore/channel_store.go:3470]: That's acceptable. I'm just looking for ways to make the code more compact.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3498]: This can be replaced by just `sq.Eq`.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3303]: This can be replaced by just `sq.Eq`.

File: store/sqlstore/channel_store.go
Comment (noxer) [inline store/sqlstore/channel_store.go:3147]: This can be replaced by just `sq.Eq`.

**Review by noxer — CHANGES_REQUESTED** [review]
Good work. Seems like I missed a few `sq.And`'s in the review. Sorry about that.

**Comment by agnivade** [comment]
@noxer - Done!

**Review by noxer — APPROVED** [review]
LGTM

**Comment by spirosoik** [comment]
@agnivade it was needing my comment :P

**Comment by agnivade** [comment]
@jgilliam17 - Need a quick check on the e2e test report please :)

**Comment by jgilliam17** [comment]
@agnivade [E2E report](https://mattermost-cypress-report.s3.amazonaws.com/169470-68b4a04-onprem-ent-server-pr-19550-mattermost/mm-ee-test68b4a04/mochawesome.html) looks good, no PR related failures when compared to latest[ master run](https://mattermost-cypress-report.s3.amazonaws.com/169344-onprem-ent-master-mm-ee-test:test/mochawesome.html)

**Review by jgilliam17 — APPROVED** [review]

**Comment by agnivade** [comment]
Brilliant, thanks as always Jelena!

**Pull request merged**
